src/libraries/System.Runtime.Serialization.BinaryFormat/src/System/Runtime/Serialization/BinaryFormat/ArraySinglePrimitiveRecord.cs

```diff
@@ -0,0 +1,212 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System.Buffers;
+using System.Buffers.Binary;
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.Globalization;
+using System.IO;
+using System.Linq;
+using System.Reflection.Metadata;
+using System.Runtime.CompilerServices;
+using System.Runtime.InteropServices;
+using System.Runtime.Serialization.BinaryFormat.Utils;
+
+namespace System.Runtime.Serialization.BinaryFormat;
+
+/// <summary>
+/// Represents a single-dimensional array of a primitive type.
+/// </summary>
+/// <remarks>
+/// ArraySinglePrimitive records are described in <see href="https://learn.microsoft.com/openspecs/windows_protocols/ms-nrbf/3a50a305-5f32-48a1-a42a-c34054db310b">[MS-NRBF] 2.4.3.3</see>.
+/// </remarks>
+internal sealed class ArraySinglePrimitiveRecord<T> : ArrayRecord<T>
+    where T : unmanaged
+{
+    private static TypeName? s_elementTypeName;
+
+    internal ArraySinglePrimitiveRecord(ArrayInfo arrayInfo, IReadOnlyList<T> values) : base(arrayInfo)
+    {
+        Values = values;
+        ValuesToRead = 0; // there is nothing to read anymore
+    }
+
+    public override RecordType RecordType => RecordType.ArraySinglePrimitive;
+
+    public override TypeName ElementTypeName
+        => s_elementTypeName ??= TypeName.Parse(typeof(T).FullName.AsSpan()).WithAssemblyName(typeof(T).GetAssemblyNameIncludingTypeForwards());
+
+    internal IReadOnlyList<T> Values { get; }
+
+    public override bool IsTypeNameMatching(Type type) => typeof(T[]) == type;
+
+    internal override bool IsElementType(Type typeElement) => typeElement == typeof(T);
+
+    /// <inheritdoc/>
+    public override T[] GetArray(bool allowNulls = true)
+        => Values is T[] array ? array : Values.ToArray();
+
+    internal override (AllowedRecordTypes allowed, PrimitiveType primitiveType) GetAllowedRecordType()
+    {
+        Debug.Fail("GetAllowedRecordType should never be called on ArraySinglePrimitiveRecord");
+        throw new InvalidOperationException();
+    }
+
+    private protected override void AddValue(object value)
+    {
+        Debug.Fail("AddValue should never be called on ArraySinglePrimitiveRecord");
+        throw new InvalidOperationException();
+    }
+
+    internal static IReadOnlyList<T> DecodePrimitiveTypes(BinaryReader reader, int count)
+    {
+        if (typeof(T) == typeof(byte) && reader.IsDataAvailable(count))
+        {
+            T[] result = (T[])(object)reader.ReadBytes(count);
+            // This might be less than the number of bytes requested if the end of the stream is reached.
+            if (result.Length != count)
+            {
+                ThrowHelper.ThrowEndOfStreamException();
+            }
+            return result;
+        }
+        else if (typeof(T) == typeof(char) && reader.IsDataAvailable(count))
```

Comment:
> So you need to double it for char, assuming there's no UTF-8 wonkiness going on.

The input is always UTF8, that is why the `* sizeof(char)` is missing on purpose (I am going to add a comment)

Reply:
If `char[]` is using UTF8, is the count the number of "code units" or "code points"?  `BinaryReader.ReadChars()` says it's returning a `char[count]`, so it thinks "code points"; but since you're using it as a length check here you are testing "code units".

The "Grinning Face" emoji ([U+1F600](https://www.compart.com/en/unicode/U+1F600)) takes 4 bytes ("code units") to encode.  So, this is possibly saying "Are there at least 1/4 the necessary bytes remaining?".

I'm personally OK with that, I'm just pointing out that this isn't a size _guarantee_.